File: cases/taint/ruby/callback-registration-positive/callback_registration.rb
class Registry
  def initialize
    @hooks = []
  end

  def register(hook)
    @hooks.push(hook)
  end

  def fire(value) # DFB-WITNESS: callback-registration-fire
    @hooks.each do |hook|
      hook.call(value)
    end
  end
end

def dfb_source # DFB-SOURCE: callback-registration-input
  "tainted"
end

def dfb_sink(value) # DFB-SINK: callback-registration-sink
end

def leak(value)
  dfb_sink(value)
end

def run
  registry = Registry.new
  registry.register(method(:leak))
  registry.fire(dfb_source)
end

File: cases/taint/python/callback-registration-positive/callback_registration.py
def dfb_source():  # DFB-SOURCE: callback-registration-input
    return "tainted"


def dfb_sink(value):  # DFB-SINK: callback-registration-sink
    pass


class Registry:
    def __init__(self):
        self.hooks = []

    def register(self, hook):
        self.hooks.append(hook)

    def fire(self, value):  # DFB-WITNESS: callback-registration-fire
        for hook in self.hooks:
            hook(value)


def leak(value):
    dfb_sink(value)


def run():
    registry = Registry()
    registry.register(leak)
    registry.fire(dfb_source())
